### PR TITLE
[SYCL][CUDA] Compile .cu files with SYCL 

### PR DESCRIPTION
**Pull request by pgorlani**
This PR allows to compile `.cu` files with SYCL, i.e.,

1. Compile `.cu` sources in order to generate `.o` files for SYCL.
2. Compile `.cu` and `.cpp` files for obtaining a SYCL executable.
3. Create a dummy (CUDA) `__host__` function for each (CUDA) `__device__` function that hasn't one.

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 1]
```diff
@@ -1640,8 +1640,17 @@ void CodeGenFunction::GenerateCode(GlobalDecl GD, llvm::Function *Fn,
            !getLangOpts().CUDAIsDevice &&
            FD->hasAttr<CUDAGlobalAttr>())
     CGM.getCUDARuntime().emitDeviceStub(*this, Args);
-  else if (isa<CXXMethodDecl>(FD) &&
-           cast<CXXMethodDecl>(FD)->isLambdaStaticInvoker()) {
+  else if (getLangOpts().CUDA && !getLangOpts().CUDAIsDevice &&
```
Comment (smanna12) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1643]: Please add a description.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 2]
```diff
@@ -2708,6 +2708,14 @@ void CodeGenModule::EmitDeferred() {
   CurDeclsToEmit.swap(DeferredDeclsToEmit);
 
   for (GlobalDecl &D : CurDeclsToEmit) {
+    if (LangOpts.CUDA && LangOpts.SYCLIsHost) {
```
Comment (smanna12) [inline clang/lib/CodeGen/CodeGenModule.cpp:2711]: Comment would be very helpful. Please add a comment.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 3]
```diff
@@ -3330,9 +3338,12 @@ void CodeGenModule::EmitGlobal(GlobalDecl GD) {
 
       // So device-only functions are the only things we skip.
       if (isa<FunctionDecl>(Global) && !Global->hasAttr<CUDAHostAttr>() &&
-          Global->hasAttr<CUDADeviceAttr>())
+          Global->hasAttr<CUDADeviceAttr>()) {
```
Comment (smanna12) [inline clang/lib/CodeGen/CodeGenModule.cpp:3341]: Same here

File: clang/test/Driver/sycl-cuda-tu-offload.cu [hunk 4]
```diff
@@ -0,0 +1,48 @@
+// RUN: %clang++ -ccc-print-phases -target x86_64-unknown-linux-gnu  -fsycl -fsycl-targets=nvptx64-nvidia-cuda  -Xsycl-target-backend  --cuda-gpu-arch=sm_80 --cuda-gpu-arch=sm_80 -c %s 2>&1 | FileCheck %s --check-prefix=DEFAULT-PHASES
+
+// DEFAULT-PHASES:         +- 0: input, "{{.*}}", cuda, (device-cuda, sm_80)
```
Comment (smanna12) [inline clang/test/Driver/sycl-cuda-tu-offload.cu:3]: Same here

**Comment by pgorlani** [comment]
Hi @smanna12, 
  thanks for your review. I added some comments, let me know if they are fine.

Basically, the main point is that in SYCL every device function has an host counterpart.

This is not always possible in CUDA, since an user could write a ``__device__`` function without a corresponding ``__host__`` one. In this case, a dummy host function that returns ``undef`` is generated.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 2 above)
Comment (pgorlani) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Yes, good point! Thank you, @premanandrao.

I added a commit in this regard.

File: clang/lib/Driver/Driver.cpp [hunk 5]
```diff
@@ -5515,6 +5579,14 @@ class OffloadingActionBuilder final {
       delete SB;
   }
 
+  void pushForeignAction(Action *A) {
```
Comment (mdtoguchi) [inline clang/lib/Driver/Driver.cpp:5582]: Could you add a comment describing what a 'foreign action' is and how it is handled (as opposed to native files)

File: clang/lib/Driver/Driver.cpp [hunk 6]
```diff
@@ -5515,6 +5579,14 @@ class OffloadingActionBuilder final {
...
+    for (auto *SB : SpecializedBuilders) {
+      if (SB->isValid()) {
+        SB->pushForeignAction(A);
+      }
```
Comment (mdtoguchi) [inline clang/lib/Driver/Driver.cpp:5586]: small nit in regards to style, I've noticed it in a couple of places.
```suggestion
      if (SB->isValid())
        SB->pushForeignAction(A);
```

File: clang/lib/Driver/Driver.cpp (hunk 5 above)
Comment (pgorlani) [reply clang/lib/Driver/Driver.cpp:5587]: Hi @mdtoguchi,
thanks for your review. I added a [commit](https://github.com/intel/llvm/pull/6395/commits/cdf91e40c206c541f6dc4d250b5e0ebf7c6033bd) with some comments, let me know how they sound.

The foreign actions are actions coming from different specialized device action builder.
Basically, we passed actions form the `CudaActionBuilder` to the `SYCLActionBuilder` making them communicating together in order to compile CUDA sources with SYCL.

**Review by premanandrao — APPROVED** [review]

**Review by smanna12 — DISMISSED** [review]

**Comment by pgorlani** [comment]
Ping @mdtoguchi

**Review by mdtoguchi — APPROVED** [review]

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 7]
```diff
@@ -1640,8 +1640,18 @@ void CodeGenFunction::GenerateCode(GlobalDecl GD, llvm::Function *Fn,
...
+           getLangOpts().SYCLIsHost && !FD->hasAttr<CUDAHostAttr>() &&
+           FD->hasAttr<CUDADeviceAttr>()) {
+    // Generate a dummy __host__ function for compiling CUDA sources in SYCL.
+    Fn->setLinkage(llvm::Function::WeakODRLinkage);
+    if (FD->getReturnType()->isVoidType())
+      Builder.CreateRetVoid();
+    else
+      Builder.CreateRet(llvm::UndefValue::get(Fn->getReturnType()));
+    return;
+  } else if (isa<CXXMethodDecl>(FD) &&
```
Comment (bader) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1653]:
```suggestion
    return;
  }
  if (isa<CXXMethodDecl>(FD) &&
```
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

**Review by bader — APPROVED** [review]
Thanks!

**Review by smanna12 — APPROVED** [review]

**Pull request merged**